includes/TripalFields/data__identifier/data__identifier_widget.inc

```diff
@@ -0,0 +1,259 @@
+   *   An associative array containing the current state of the form.
+   * @param $langcode
+   *   The language associated with $items.
+   * @param $items
+   *   Array of default values for this field.
+   * @param $delta
+   *   The order of this item in the array of subelements (0, 1, 2, etc).
+   * @param $element
+   * A form element array containing basic properties for the widget:
+   *  - #entity_type: The name of the entity the field is attached to.
+   *  - #bundle: The name of the field bundle the field is contained in.
+   *  - #field_name: The name of the field.
+   *  - #language: The language the field is being edited in.
+   *  - #field_parents: The 'parents' space for the field in the form. Most
+   *    widgets can simply overlook this property. This identifies the location
+   *    where the field values are placed within $form_state['values'], and is
+   *    used to access processing information for the field through the
+   *    field_form_get_state() and field_form_set_state() functions.
+   *  - #columns: A list of field storage columns of the field.
+   *  - #title: The sanitized element label for the field instance, ready for
+   *    output.
+   *  - #description: The sanitized element description for the field instance,
+   *    ready for output.
+   *  - #required: A Boolean indicating whether the element value is required;
+   *    for required multiple value fields, only the first widget's values are
+   *    required.
+   *  - #delta: The order of this item in the array of subelements; see
+   *    $delta above
+   */
+  public function form(&$widget, &$form, &$form_state, $langcode, $items, $delta, $element) {
+    // This widget applies only to type data identifier although it is attached to text
+    // field which is a general field type. An warning if this is applied to other text field.
+
+    if ($element['#field_name'] == 'data__identifier' && strtolower($element['#title']) == 'identifier') {
+      // In identifier field.
+
+      // Text field. In either cases this field must be disabled from user.
+      $field_name = $this->field['field_name'];
+      $field_table = $this->instance['settings']['chado_table'];
+      $linker_field = 'chado-' . $field_table . '__uniquename';
+
+      // When interface to set prefix:
+      // Notice for admin where they can alter the prefix use to formulate uniquename.
+      // @see hook_menu().
+      $link = l('Update Prefix', 'admin/tripal/extension/trpfancy_fields/prefix');
+      tripal_set_message('Administrators, you can set prefix used in uniquename field in ' . $link . '.', TRIPAL_INFO, array('return_html' => FALSE));
+
+      // Dertermine if process is Create or Update. To test: check if chado_record_id
+      // were set to an id number.
+      if (isset($element['#entity']->chado_record_id) && $element['#entity']->chado_record_id > 0) {
+        // Update here.
+
+        $unique_name = array_key_exists($delta, $items) ? $items[$delta]['value'] : '';
```

Comment:
On update, uniquename should be set to the prefix + ID rather then the existing uniquename. Right now it's confusing b/c I see the old uniquename in the textfield but then on save it changes.